bigquery/google/cloud/bigquery/table.py

```diff
@@ -616,41 +619,58 @@ def _set_properties(self, api_response):
             cleaned['expirationTime'] = float(cleaned['expirationTime'])
         self._properties.update(cleaned)
 
-    def _build_resource(self):
+    def _populate_expires_resource(self, resource):
+        value = _millis_from_datetime(self.expires)
+        resource['expirationTime'] = value
+
+    def _populate_partitioning_type_resource(self, resource):
+        resource['timePartitioning'] = self._properties['timePartitioning']
+
+    def _populate_view_use_legacy_sql_resource(self, resource):
+        if 'view' not in resource:
+            resource['view'] = {}
+        resource['view']['useLegacySql'] = self.view_use_legacy_sql
+
+    def _populate_view_query_resource(self, resource):
+        if 'view' not in resource:
+            resource['view'] = {}
+        resource['view']['query'] = self.view_query
+
+    def _populate_schema_resource(self, resource):
+        resource['schema'] = {'fields': _build_schema_resource(self._schema)}
+
+    def _build_resource(self, filter_fields=[]):
         """Generate a resource for ``create`` or ``update``."""
         resource = {
             'tableReference': {
                 'projectId': self._project,
                 'datasetId': self._dataset_id,
                 'tableId': self.table_id},
         }
-        if self.description is not None:
-            resource['description'] = self.description
-
-        if self.expires is not None:
-            value = _millis_from_datetime(self.expires)
-            resource['expirationTime'] = value
-
-        if self.friendly_name is not None:
-            resource['friendlyName'] = self.friendly_name
-
-        if self.location is not None:
-            resource['location'] = self.location
-
-        if self.partitioning_type is not None:
-            resource['timePartitioning'] = self._properties['timePartitioning']
-
-        if self.view_query is not None:
-            view = resource['view'] = {}
-            view['query'] = self.view_query
-            if self.view_use_legacy_sql is not None:
-                view['useLegacySql'] = self.view_use_legacy_sql
-
-        if self._schema:
-            resource['schema'] = {
-                'fields': _build_schema_resource(self._schema)
-            }
-
+        custom_resource_fields = {
+            'expires': self._populate_expires_resource,
+            'partitioning_type': self._populate_partitioning_type_resource,
+            'view_query': self._populate_view_query_resource,
+            'view_use_legacy_sql': self._populate_view_use_legacy_sql_resource,
+            'schema': self._populate_schema_resource
+        }
+        all_fields = [
+            'description', 'friendly_name', 'expires', 'location',
+            'partitioning_type', 'view_use_legacy_sql', 'view_query', 'schema'
+        ]
+        for f in all_fields:
+            if filter_fields and f not in filter_fields:
```

Comment:
It looks like you intend that if `filter_fields` is None or [], then all fields should be updated. Two issues I have with that:
1. If the caller of update_table accidentally passes in [], they'll overwrite their entire table. Let's avoid that case.
2. The logic on this line is a little tricky to follow.

One solution: in update_table, check that `properties` is a sequence. Here, make the default None, and near the top say
```
if filter_fields is None:
    filter_fields = all_fields
```

Reply:
If the caller of update_table accidentally passes in [], they'll only update the fields that are not None or empty. Fields that are not filtered in that are None or empty are skipped in the next conditional on line 664. What is the desired behavior when the user accidentally passes in []?

With the current code, if the user wants an unfiltered resource, it will not include any properties with values that are None or empty. If there are filters, the resource will include the property regardless of its value. If I set `filter_fields` equal to `all_fields` when there aren't any filters, there's no longer a way to determine whether a field was specifically meant to be filtered in regardless of value.

Reply:
> What is the desired behavior when the user accidentally passes in []?

The call is a no-op.

Regarding the second point: I think this function is only called in two places: 

1. `create_table`,  in which case we want to send all the fields (or maybe all the non-empty ones; I'm not sure it matters, but in any case, only one of those two choices)
2. `update_table`, where the user passes in the list of fields and we send exactly those, regardless of their value.

I don't think there is a third case.

Reply:
@tswast had an idea for how to fix this. With https://github.com/GoogleCloudPlatform/google-cloud-python/pull/4076/commits/19af93abcde93af803b07f6f3ebe468e30c01c11 I updated `create_table()` to send in `all_fields` as the `filter_fields`. Then `_build_resource()` only loops through the filter_fields. This removed the messy logic. After the resource is built, `create_table()` removes any fields that are None before sending the API request. 

As for `update_table()`, with the way it is now set up, nothing will happen if a user passes in [] because the resource builder only loops through the `filter_fields`.

Reply:
SGTM.